internal/controller/elementalhost_controller.go

```diff
@@ -154,6 +191,31 @@ func (r *ElementalHostReconciler) reconcileNormal(ctx context.Context, host *inf
 	return ctrl.Result{}, nil
 }
 
+func (r *ElementalHostReconciler) reconcileOSVersionManagement(ctx context.Context, host *infrastructurev1beta1.ElementalHost) error {
+	logger := log.FromContext(ctx).
+		WithValues(ilog.KeyNamespace, host.Namespace).
+		WithValues(ilog.KeyElementalHost, host.Name).
+		WithValues(ilog.KeyElementalMachine, host.Spec.MachineRef.Name)
+	logger.Info("Reconciling OSVersionManagement from associated ElementalMachine")
+	machine := &infrastructurev1beta1.ElementalMachine{}
+	err := r.Client.Get(ctx, client.ObjectKey{Namespace: host.Spec.MachineRef.Namespace, Name: host.Spec.MachineRef.Name}, machine)
+	// If the ElementalMachine was not found, assume it was deleted.
+	// Best thing we can do is to trigger reset by deleting this ElementalHost.
+	// This has chances to happen if an ElementalMachine is deleted and the finalizer is removed.
+	if apierrors.IsNotFound(err) {
+		logger.Error(err, "Associated ElementalMachine is not found. Triggering Host reset.")
+		now := metav1.Now()
+		host.SetDeletionTimestamp(&now)
```

Comment:
This probably needs to be checked before being set.
Some race condition can happen otherwise:
```
2024-04-19T14:28:09Z	ERROR	Associated ElementalMachine is not found. Triggering Host reset.	{"controller": "elementalhost", "controllerGroup": "infrastructure.cluster.x-k8s.io", "controllerKind": "ElementalHost", "ElementalHost": {"name":"m-25e45776-236e-4458-a749-9b8ae67afa5c","namespace":"default"}, "namespace": "default", "name": "m-25e45776-236e-4458-a749-9b8ae67afa5c", "reconcileID": "0fc11e99-a591-4e8a-8608-ec5d2e1b56cd", "Namespace": "default", "ElementalHost": "m-25e45776-236e-4458-a749-9b8ae67afa5c", "ElementalMachine": "rke2-control-plane-rqtjh", "error": "ElementalMachine.infrastructure.cluster.x-k8s.io \"rke2-control-plane-rqtjh\" not found"}
github.com/rancher-sandbox/cluster-api-provider-elemental/internal/controller.(*ElementalHostReconciler).reconcileOSVersionManagement
	/workspace/internal/controller/elementalhost_controller.go:206
github.com/rancher-sandbox/cluster-api-provider-elemental/internal/controller.(*ElementalHostReconciler).reconcileNormal
	/workspace/internal/controller/elementalhost_controller.go:163
github.com/rancher-sandbox/cluster-api-provider-elemental/internal/controller.(*ElementalHostReconciler).Reconcile
	/workspace/internal/controller/elementalhost_controller.go:137
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:119
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:316
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:227
2024-04-19T14:28:09Z	ERROR	Reconciler error	{"controller": "elementalhost", "controllerGroup": "infrastructure.cluster.x-k8s.io", "controllerKind": "ElementalHost", "ElementalHost": {"name":"m-25e45776-236e-4458-a749-9b8ae67afa5c","namespace":"default"}, "namespace": "default", "name": "m-25e45776-236e-4458-a749-9b8ae67afa5c", "reconcileID": "0fc11e99-a591-4e8a-8608-ec5d2e1b56cd", "error": "reconciling ElementalHost: reconciling OSVersionManagement: fetching non-existing ElementalMachine 'rke2-control-plane-rqtjh': ElementalMachine.infrastructure.cluster.x-k8s.io \"rke2-control-plane-rqtjh\" not found\npatching ElementalHost: ElementalHost.infrastructure.cluster.x-k8s.io \"m-25e45776-236e-4458-a749-9b8ae67afa5c\" is invalid: metadata.deletionTimestamp: Invalid value: 2024-04-19 14:28:09 +0000 UTC: field is immutable"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:329
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:266
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:227
```